glean_parser/util.py

```diff
@@ -66,13 +66,35 @@ def remove_implicit_resolver(cls, tag_to_remove):
 _NoDatesSafeLoader.remove_implicit_resolver("tag:yaml.org,2002:timestamp")
 
 
+class SafeLineLoader(_NoDatesSafeLoader):
+    """
+    Map line number to yaml nodes.
+    Source: https://stackoverflow.com/questions/13319067/parsing-yaml-return-with-line-number
+    """
+
+    def construct_mapping(self, node, deep=False):
+        mapping = super(SafeLineLoader, self).construct_mapping(
+            node, deep=deep)
+        line_number = {"line": str(node.start_mark.line)}
+
+        for value in node.value:
+            if value[0].value != "$schema":
+                # pings
+                if node.start_mark.column == 2 and value[0].value == "description":
+                    mapping['defined_in'] = line_number
+                # metrics
+                if node.start_mark.column == 4 and value[0].value == "type":
+                    mapping['defined_in'] = line_number
```

Comment:
The logic used here is admittedly hacky, but right now I can't think of a better way

Reply:
Yeah -- this stuff is always hard and hacky to do.  Rather than focusing on the specific fields `description` and `type` and column numbers, what if we just add the first line we see to every mapping, something like:

```
mapping['defined_in`] = min(line_number, mapping.get('defined_in', sys.maxint))
```

Reply:
@mdboom The reason why the columns number, and value checking of “description” or “type” were used is because due to the differences in the yaml file structures for pings and metrics, there’s only a few very specific places where we can map `defined_in` to:

- pings.yaml: we want to map only when conditions are node `column == 2` and `value == “description”` to make sure that a) it is a ping we’re mapping to, and b) `defined_in` will be at the same level as the other ping properties. 

- metrics.yaml: we don’t want to map `defined_in` at `column == 2` because it will map it at the same node level as the other metrics, which `defined_in` then becomes another metric instead of just a metric property. `column == 4` and `value == “type”` will make sure that `defined_in` will be mapped at the metric property level. These 2 conditions also help to avoid mapping to nodes like [this one](https://github.com/mozilla/glean/blob/1da5cdf7300e814bd5993652b0a60a11c45c5ed6/glean-core/pings.yaml#L30) whose column also starts at 4.

I hope that made sense. I did try what you suggested, but I’ve found that without being very specific with the mapping condition it will cause errors.